### ISSUE TITLE
.PNG Tab Icons on iOS aren't scaled properly

### ISSUE THREAD

**Issue by Core121** [dotnet/maui#8609]
### Description

Tab Icons that are of type .png do not resize on iOS.

### Steps to Reproduce

1. New .NET MAUI App
2. Add TabBar with .png icons in AppShell.xaml
3. Launch app on iOS
4. View the following screen
![image](https://user-images.githubusercontent.com/42871786/178002561-3a231fa2-1c80-44fe-a73e-036b558acaac.png)

View from Android 12.0 - API 31
![image](https://user-images.githubusercontent.com/42871786/178003471-fe22756f-f971-4cc4-9821-357172c8c523.png)

Code:
[TabIconTest.zip](https://github.com/dotnet/maui/files/9072502/TabIconTest.zip)

### Version with bug

6.0.400 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

iOS

### Affected platform versions

iOS 15.5

**Comment by edgiardina** [comment]
Also experiencing this; my icons are not resized on the tabbar. and the 'More' page for items that don't fit on the bottom tabar, the icons also render very very large.

https://github.com/edgiardina/IfpaMaui

**Comment by ignaciofuentes** [comment]
Have you tried with svg? it's the same problem. Works fine on Android. On iOS the icons look huge

**Comment by Mous625** [comment]
Can the Maui team provide some insight into when this will be fixed? Critical issues such as this shouldn't take so long to be fixed.

**Comment by edgiardina** [comment]
I don't think we should be making demands of the team. There's lot of issues to address. All we can do is emphasize how severe the issue is. It's the biggest issue preventing us from migrating from Xamarin Forms at the moment. ![image](https://user-images.githubusercontent.com/3627193/187558308-01712733-f795-4893-ac42-70452215fcce.png)

I own a small private app I want to migrate as well as running an enterprise app team. Both of those Xamarin Forms apps can't be migrated due to critical flaws in iOS at the moment (performance, image scaling, iOS internals crashing). Unfortunately I can't contribute to the MAUI main code base as we're busy shipping our own apps but I hope iOS gets prioritized. I am happy to provide my sample repo which is a real world migration I've done mostly by hand.

https://github.com/edgiardina/IfpaMaui

![image](https://user-images.githubusercontent.com/3627193/187558915-9e888e7d-23be-4576-8902-a2533e536b59.png)

**Comment by InetJustin** [comment]
Joining in the chorus, I'm using .svg images and it's the same thing, the icons are huge.

**Comment by MGohil** [comment]
As a workaround, for .png icons, you can resize those icons to 30 pixels and they will look good in both platforms.

![image](https://user-images.githubusercontent.com/16605284/192452017-3f198c51-2f65-4bbd-9229-531b82ec0a57.png)

Result in iOS:
![image](https://user-images.githubusercontent.com/16605284/192452254-8631a030-dc61-4f82-9fe4-d5dcda47f499.png)

**Comment by InetJustin** [comment]
@MGohil I appreciate your workaround, I've tried using only 30x30 .png files and the icons are still showing up really large in the bottom tab for some reason. If I use the same file for the flyout icon, it looks perfect. I've also noticed that if I change the phone to be horizontal, the icon in the bottom tab will look correct, it's only in the vertical position where it looks to be really huge and obscures the bottom left corner of the screen. Any advice you or might have would be greatly appreciated.

**Comment by Mataboge** [comment]
I can confirm issue using svg

**Comment by imyzlv** [comment]
The workaround I've found is to actually scale the svg to 30x30 and then it seems to work fine.

**Comment by Mataboge** [comment]
@imyzlv how do you do that?

**Comment by imyzlv** [comment]
Valid point.
However, I used some online service for it and then it at least works.

**Comment by mattleibow** [comment]
Are you able to use the BaseSize on the relevant images?


For example, in the linked repro, you can just "update" the images with a base size:

```xml
<MauiImage Update="Resources\Images\groceries.png" BaseSize="20,20" />
<MauiImage Update="Resources\Images\shopping_cart.png" BaseSize="20,20" />
```

You can also use a sub-folder or a seprate folder and just do a bulk update/include. For example, if you have a new folder "TabIcons" inside "Images" (`Resources\Images\TabIcons\*`):


```xml
<MauiImage Update="Resources\Images\TabIcons\**" BaseSize="20,20" />
```

**Comment by vhugogarcia** [comment]
> Are you able to use the BaseSize on the relevant images?
> 
> For example, in the linked repro, you can just "update" the images with a base size:
> 
> ```
> <MauiImage Update="Resources\Images\groceries.png" BaseSize="20,20" />
> <MauiImage Update="Resources\Images\shopping_cart.png" BaseSize="20,20" />
> ```
> 
> You can also use a sub-folder or a seprate folder and just do a bulk update/include. For example, if you have a new folder "TabIcons" inside "Images" (`Resources\Images\TabIcons\*`):
> 
> ```
> <MauiImage Update="Resources\Images\TabIcons\**" BaseSize="20,20" />
> ```

The idea of setting the BaseSize worked very good for me on iOS and Android, I was facing this issue as well on RC2 .NET 7. Thanks for sharing this workaround @mattleibow .

**Comment by InetJustin** [comment]
Can confirm this works, thank you @mattleibow 

`<MauiImage Update="Resources\Images\groceries.png" BaseSize="20,20" />`

**Comment by Syed-RI** [comment]
Does the base size workaround works on PNG as well?

**Comment by mattleibow** [comment]
Setting the BaseSize is not 100% a workaround but rather a feature.

The iOS OS has a bug/feature/support for _any_ image size so providing a 100x100 image with just be used without question. Most other OSes are sane and realize that the tab icon should fit in the tab...

The correct way to do icons in iOS is to create smaller images - like mentioned here https://github.com/dotnet/maui/issues/8609#issuecomment-1282599646

And the correct way to do these tab icons in MAUI is to also set a base size.

And this is where doing things automagically is really deceiving: if you have a large image - say 1024x1024 (because you got that from a designer or that is just the size the drawing app used) - then you actually force the OS to load that HUGE image and then scale it down to 20x20 at runtime. This definitely impacts startup performance - even in release builds.

Your app will have several large image files (increasing file size) and then the app has to load those images (long load times) and then has to resize those image (even more load time).

I am going to close this issue "as designed" because the BaseSize attribute is specifically what it was designed for. Maui has no way of knowing what the image is going to be used for and thus it is up to you to create images or set scaling so that the tooling can generate the correct images.

**Comment by edgiardina** [comment]
I don't want to get into a fight about this, but ... we didn't have to set basesize in Xamarin Forms. My images are all tiny because they're SVGs, but because I sourced them from various places, they all have an SVG internal size or whatever, leading to the confusion. I think a lot of new developers aren't going to be aware of BaseSize and run into this issue. Can I ask what changed between XF and MAUI that image size matters for the tab icons?

Plus, is this bug still a factor? would basesize fix it if the image is an SVG?

https://github.com/dotnet/maui/issues/9716

FYI, the workaround on my issue was finding a bulk SVG resizer online and running all my assets through that.

**Comment by mattleibow** [comment]
@edgiardina you are correct that new devs will hit this. There is some docs: https://learn.microsoft.com/dotnet/maui/user-interface/graphics/images#resize-an-image

But I will open a new issue to discuss the best way to solve this for new devs: https://github.com/dotnet/maui/issues/11099

With Xamarin.Forms, I do not believe there was a feature that took svgs and converted them to pngs. Are you able to attach a repro to the new issue? https://github.com/dotnet/maui/issues/11099

And yes, BaseSize should work on SVGs - in fact they were first implemented for SVGs. There have been several bug fixes so the latest builds of MAUI should be working.

**Comment by mattleibow** [comment]
Marking this as a duplicate of https://github.com/dotnet/maui/issues/11099 so the issues prioritizer can link this.

Duplicate of #11099

**Comment by mattleibow** [comment]
Duplicate of #11099

**Comment by sskrishnan9582** [comment]
Had the same problem. Resizer does not resizing the svg for iOS devices. I had to manually set to 30X30 for icons to show correctly in the tabs. Hope it will help someone.

**Comment by tuliocll** [comment]
It's a good idea add this issue on the [tutorial page](https://learn.microsoft.com/pt-br/dotnet/maui/tutorials/notes-app/?source=recommendations&view=net-maui-7.0&tutorial-step=2) as a tip for iOS.

**Comment by adegeo** [comment]
@mattleibow How is this a duplicate? Why is the iOS layout allowed to have the image stretch beyond the limits of the tab bar area? Shouldn't it be confined?

**Comment by mattleibow** [comment]
I think @adegeo is correct here. Sure, resizetizer is not supposed to resize the image to the tab size as it has no way of knowing. And sure, apple is wrong.

But maybe we can do something with the tabs? @PureWeen not sure if there are things to force ios to be reasonable?

**Comment by josueeap** [comment]
I added my 31x31 svg images and then in the icon property of the tab I just added the name of the image without the .svg extension.

That worked for me.

`    <TabBar>
       <Tab Title="API"
            Icon="tab_report">
           <ShellContent
                ContentTemplate="{DataTemplate local:MainPage}"
                Route="MainPage"/>
       </Tab>
       <Tab Title="Timer"
            Icon="tabl_timer">
           <ShellContent
                ContentTemplate="{DataTemplate local:PageTimer}"
                Route="PageTimer"/>
       </Tab>
       <Tab Title="Settings"
            Icon="tab_settings">
           <ShellContent
                ContentTemplate="{DataTemplate local:PageSettings}"
                Route="PageSettings"/>
       </Tab>
    </TabBar>`

![IMG_3272](https://user-images.githubusercontent.com/13387481/222340143-3f1e618d-fdb5-4a9d-b595-57230a7c636d.jpg)

**Comment by vhugogarcia** [comment]
@josueeap awesome. How can you set the size of an image svg? Could you share a file sample? Please. Also, does it work on iOS and Android?

**Comment by josueeap** [comment]
@vhugogarcia Sure I added an image svg example, yes it works on iOS and Android, I tried on physical devices.

I hope it helps.

I used [this online tool](https://www.iloveimg.com/resize-image/resize-svg) to resize svg images.

**SVG example:**
![tab_settings](https://user-images.githubusercontent.com/13387481/222524493-3410fb57-f757-49c3-9101-073d705a0122.svg)

**Xiamoi Android 13.0 - API 33**
![Android](https://user-images.githubusercontent.com/13387481/222523587-599aebf3-fdf6-4236-92c9-d1c0c7c61edb.jpg)

**iPhone 14 pro - iOS 16.2**
![iPhone14](https://user-images.githubusercontent.com/13387481/222524091-40e50363-351f-4139-94f3-ee63f089b56a.PNG)

**iPhone 6S - iOS 15.7.2**
![iPhone6](https://user-images.githubusercontent.com/13387481/222524232-40617f11-1d21-4781-9731-48d7b8413f0e.PNG)

**Comment by vhugogarcia** [comment]
Thanks @josueeap for the detailed response. I will take a look at this approach. Thanks again!

**Comment by mattleibow** [comment]
Does the BaseSize attribute not work?

**Comment by mattleibow** [comment]
Looking at things, the best way is to set the `<MauiImage Include="..." BaseSize="32,32" />` as this will generate nice small images that are fast to load and no additional processing required.

However, since this cannot be guaranteed, we can also make sure the icon size is correct on the tabs using `ImageInsets`:

```cs
var tabBarItem = controller.TabBarItem;

var tabItemSize = new CGSize(32, 32);

var actualSize = tabBarItem.SelectedImage?.Size ?? tabBarItem.Image?.Size ?? CGSize.Empty;

var maxW = Math.Max(actualSize.Width, tabItemSize.Width);
var minW = Math.Min(actualSize.Width, tabItemSize.Width);
var maxH = Math.Max(actualSize.Height, tabItemSize.Height);
var minH = Math.Min(actualSize.Height, tabItemSize.Height);

var insets = new UIEdgeInsets(
    (nfloat)((maxH - minH) / 2.0),
    (nfloat)((maxW - minW) / 2.0),
    (nfloat)((maxH - minH) / 2.0),
    (nfloat)((maxW - minW) / 2.0f));

tabBarItem.ImageInsets = insets;
```

**Comment by mattleibow** [comment]
We still can't really force a size because Apple recommends different sizes based on shape: https://developer.apple.com/design/human-interface-guidelines/components/navigation-and-search/tab-bars/

Is there a reason why BaseSize is not working?

**Comment by Syed-RI** [comment]
Setting BaseSize works for me. May be we should dedicate a doc section for this?

**Comment by mattleibow** [comment]
We have a tiny bit of comment here: https://learn.microsoft.com/en-us/dotnet/maui/user-interface/images/images?view=net-maui-7.0

@davidbritch  maybe the docs need to be more explicit? For example, on the tab bar section we should have a section the talks about performance and best practices? https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/shell/tabs?view=net-maui-7.0

We can have things about that you need to make sure that the images you are loading are the correct size for the use case and links to the docs that describe sizes and styles. For example, Apple has different recommendations if the image is square, round, portrait or landscape. All the sizes are slightly different. And even though Android and Windows look correct, they are in fact loading a huge image and resizing it after the fact.

This is not just for tabs, but also menus, toolbars, and other icons. In fact, general images. Maybe we need an image section in the docs that just talk about how to do sizing correctly. And maybe we need a better UI for all this in the IDE, but we can start with docs so devs can make better decisions about their images.

**Comment by Syed-RI** [comment]
A huge thumbs up on @mattleibow comments above!

**Comment by mattleibow** [comment]
Closing this for a few reasons:
 - this is working as designed
 - large images on screen are a result of large images in files
 - the correct way to scale images is to either create svgs with the exact bounds or use the `BaseSize` attribute
 - we have an issue tracking the docs improvement: https://github.com/dotnet/docs-maui/issues/1464
 - we have an issue tracking an idea to maybe add a "use case" attribute for images so you don't have to figure out the size, you just say what you want to use it for
